generative_ai/function_calling.py

```diff
@@ -0,0 +1,58 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# [START aiplatform_function_calling]
+from vertexai.preview.generative_models import (
+    FunctionDeclaration,
+    GenerativeModel,
+    Tool,
+)
+
+
+def generate_function_call(prompt: str) -> str:
+    # Load the Vertex AI Gemini API to use function calling
+    model = GenerativeModel("gemini-pro")
+
+    # Specify a function declaration and parameters for an API request
+    get_current_weather_func = FunctionDeclaration(
+        name="get_current_weather",
+        description="Get the current weather in a given location",
+        # Function parameters are specified in OpenAPI JSON schema format
+        parameters={
+            "type": "object",
+            "properties": {"location": {"type": "string", "description": "Location"}},
+        },
+    )
+
+    # Define a tool that includes the above get_current_weather_func
+    weather_tool = Tool(
+        function_declarations=[get_current_weather_func],
+    )
+
+    # Prompt to ask the model about weather, which will invoke the Tool
+    prompt = prompt
+
+    # Instruct the model to generate content using the Tool that you just created:
+    response = model.generate_content(
+        prompt,
+        generation_config={"temperature": 0},
+        tools=[weather_tool],
+    )
+
+    return str(response)
```

Comment:
```suggestion
    print(response)
    return response.text
```

Reply:
This is using the function calling feature that just landed in the latest `google-cloud-aiplatform` package earlier today, and for some reason when I ask for `response.text` or other variants/attributes that I've tried, I get:

```
In [7]: response.text
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[7], line 1
----> 1 response.text

File ~/miniforge3/lib/python3.10/site-packages/vertexai/generative_models/_generative_models.py:1309, in GenerationResponse.text(self)
   1307 if len(self.candidates) > 1:
   1308     raise ValueError("Multiple candidates are not supported")
-> 1309 return self.candidates[0].content.parts[0].text

File ~/miniforge3/lib/python3.10/site-packages/vertexai/generative_models/_generative_models.py:1505, in Part.text(self)
   1502 @property
   1503 def text(self) -> str:
   1504     if "text" not in self._raw_part:
-> 1505         raise ValueError("Part has no text.")
   1506     return self._raw_part.text

ValueError: Part has no text.
```

Whereas getting the string representation of the response outputs the full contents that I can test against and users can work with.

Reply:
understood! Let me drop this.